binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java

```diff
@@ -49,118 +48,142 @@ abstract class SerdeResolverUtils {
-			// If the type is Object, then skip assigning the JsonSerde and let the fallback mechanism takes precedence.
-			if (!genericRawClazz.isAssignableFrom((Object.class))) {
-				serde = new JsonSerde(genericRawClazz);
-			}
+		if (UUID.class.isAssignableFrom(genericRawClazz)) {
+			return Serdes.UUID();
 		}
-		return serde;
+		return null;
 	}
 
 	private static boolean isSerdeFromStandardDefaults(Serde<?> serde) {
-		if (serde != null) {
-			if (Number.class.isAssignableFrom(serde.getClass())) {
-				return true;
-			}
-			else if (Serdes.ByteArray().getClass().isAssignableFrom(serde.getClass())) {
-				return true;
-			}
-			else if (Serdes.String().getClass().isAssignableFrom(serde.getClass())) {
-				return true;
-			}
-			else if (Serdes.UUID().getClass().isAssignableFrom(serde.getClass())) {
-				return true;
-			}
+		if (serde == null) {
+			return false;
+		}
+		if (Serdes.Integer().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		if (Serdes.Long().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		if (Serdes.Short().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		if (Serdes.Double().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		if (Serdes.Float().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		else if (Serdes.ByteArray().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		else if (Serdes.String().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		else if (Serdes.UUID().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
 		}
 		return false;
 	}
 
 	/**
-	 * Find all {@link Serde} beans that are assignable from {@code targetType}.
+	 * Find all {@link Serde} beans that can be used for {@code targetType}.
 	 *
 	 * @param context the application context
 	 * @param targetType the target type the serdes are being matched for
-	 * @return list of matching serdes order by most specific match, or an empty list if no matches found
+	 * @return list of matching serdes ordered by most specific match, or an empty list if no matches found
 	 */
 	static List<Serde<?>> findMatchingSerdes(ConfigurableApplicationContext context, ResolvableType targetType) {
 		// We don't attempt to find a matching Serde for type '?'
 		if (targetType.getRawClass() == null) {
 			return Collections.emptyList();
 		}
 		List<SerdeWithSpecificityScore> matchingSerdes = new ArrayList<>();
-
-		context.getBeansOfType(Serde.class).forEach((beanName, serdeBean) -> {
-			final Class<?> beanConfigClass = ClassUtils.resolveClassName(((AnnotatedBeanDefinition)
-					context.getBeanFactory().getBeanDefinition(beanName))
-					.getMetadata().getClassName(),
-				ClassUtils.getDefaultClassLoader());
+		ResolvableType serdeType = ResolvableType.forClassWithGenerics(Serde.class, targetType);
+		String[] serdeBeanNames = BeanFactoryUtils.beanNamesForTypeIncludingAncestors(context.getBeanFactory(),
```

Comment:
@artembilan I was not able to use `context.getBeanProvider` as it did not handle sorting (it does respect `@Order` but that will not work here as the order is dependent on the target type, not the bean definition) - it otherwise returns them in order they are defined in the config class. However, I was able to use similar `BeanFactoryUtils.beanNamesForTypeIncludingAncestors` to get the proper bean names and then grab the bean and bean definition so I can gets its actual type and sort by it. 

This let me simplify things 2 ways:
1. no reflection on the config class to get the return type of the bean method (I use the bean definition for this now)
2. no need to be the final arbiter w/ manual call of `serdeBeanGeneric.isAssignableFrom` (the `beanNamesForTypeIncludingAncestors` handles all of that internally now).